File: src/jkorn/pvpcore/world/PvPCWorld.php
<?php

declare(strict_types=1);

namespace jkorn\pvpcore\world;

use jkorn\pvpcore\utils\IKBObject;
use jkorn\pvpcore\utils\PvPCKnockback;
use jkorn\pvpcore\utils\IExportedValue;
use jkorn\pvpcore\utils\Utils;
use pocketmine\Player;
use pocketmine\Server;
use pocketmine\level\Level;

/**
 * Created by PhpStorm.
 * User: jkorn2324
 * Date: 2019-04-05
 * Time: 11:17
 */
class PvPCWorld implements IKBObject
{

    /** @var Level|null */
    private $level;

    /** @var bool */
    private $customKb;

    /** @var PvPCKnockback */
    private $knockbackInfo;
    /** @var string */
    private $localizedLevel;

    /**
     * PvPCWorld constructor.
     * @param string $lvl
     * @param bool $kb
     * @param PvPCKnockback $knockback
     */
    public function __construct(string $lvl, bool $kb, PvPCKnockback $knockback)
    {
        $this->customKb = $kb;
        $this->localizedLevel = $lvl;
        $this->level = Server::getInstance()->getLevelByName($lvl);
        $this->knockbackInfo = $knockback;
    }

    /**
     * @return string
     *
     * Gets the localized name of the level.
     */
    public function getLocalizedLevel(): string
    {
        return $this->localizedLevel;
    }

    /**
     * @return Level|null
     *
     * Gets the level in the world.
     */
    public function getLevel(): ?Level
    {
        return $this->level;
    }

    /**
     * @param Level $level
     *
     * Called to reload the level information.
     */
    public function setLevel(Level $level): void
    {
        $this->localizedLevel = $level->getName();
        $this->level = $level;
    }


    /**
     * @return bool
     *
     * Determines if the KB is enabled or not.
     */
    public function isKBEnabled(): bool
    {
        return $this->customKb;
    }

    /**
     * @param bool $b
     *
     * Sets the custom kb.
     */
    public function setKBEnabled(bool $b): void
    {
        $this->customKb = $b;
    }

    /**
     * @return PvPCKnockback
     *
     * Gets
     */
    public function getKnockback(): PvPCKnockback
    {
        return $this->knockbackInfo;
    }

    /**
     * @return array
     *
     * Converts the world to an array.
     */
    public function toArray(): array
    {
        return [
            "kbEnabled" => $this->customKb,
            "kbInfo" => $this->getKnockback()->toArray()
        ];
    }

    /**
     * @param $object
     * @return bool
     *
     * Determines if the objects are equivalent.
     */
    public function equals($object): bool
    {
        if ($object instanceof PvPCWorld) {
            $level = $object->getLevel();
            if ($level instanceof Level && $this->level instanceof Level) {
                return Utils::areLevelsEqual($this->level, $level);
            }

            return $this->knockbackInfo->equals($object->getKnockback());
        }

        return false;
    }

    /**
     * @param Player $player1
     * @param Player $player2
     * @return bool
     *
     * Determines if the players can use the custom knockback.
     */
    public function canUseKnocback(Player $player1, Player $player2): bool
    {
        if (!$this->level instanceof Level || !$this->customKb) {
            return false;
        }

        return Utils::areLevelsEqual($player1->getLevel(), $this->level)
            && Utils::areLevelsEqual($player2->getLevel(), $this->level);
    }

    /**
     * @param string $worldName
     * @param array $data
     * @return PvPCWorld|null
     *
     * Decodes the data & turns it into a PvPCWorld object according to the new format.
     */
    public static function decode(string $worldName, array $data): ?PvPCWorld
    {
        if (isset($data["kbEnabled"], $data["kbInfo"])) {
            $kbEnabled = (bool)$data["kbEnabled"];
            $knockback = PvPCKnockback::decode($data["kbInfo"]) ?? new PvPCKnockback();
            return new PvPCWorld(
                $worldName,
                $kbEnabled,
                $knockback
            );
        }

        return null;
    }

    /**
     * @param string $worldName
     * @param $data
     * @return PvPCWorld|null
     *
     * Decodes the data and turns it into a PvPCWorld object based on the old format.
     */
    public static function decodeLegacy(string $worldName, $data): ?PvPCWorld
    {
        $customKB = true;
        $attackDelay = 10;
        $knockback = 0.4;

        if (isset($data["customKb"])) {
            $customKB = $data["customKb"];
        }

        if (isset($data["attack-delay"])) {
            $attackDelay = (int)$data["attack-delay"];
        }

        if (isset($data["knockback"])) {
            $knockback = (float)$data["knockback"];
        }

        return new PvPCWorld(
            $worldName,
            $customKB,
            new PvPCKnockback($knockback, $knockback, $attackDelay)
        );
    }
}

File: Nukkit/src/pvpcore/worlds/PvPCWorld.java
package pvpcore.worlds;

import cn.nukkit.Player;
import cn.nukkit.Server;
import cn.nukkit.level.Level;
import org.json.simple.JSONObject;
import pvpcore.utils.IKnockbackObject;
import pvpcore.utils.PvPCKnockback;
import pvpcore.utils.Utils;

import java.util.HashMap;

public class PvPCWorld implements IKnockbackObject {

    /** The knockback input. */
    private PvPCKnockback knockbackInfo;
    /** The level name. */
    private String levelName;
    /** The corresponding level. */
    private Level level;
    /** Determines whether the KB is enabled. */
    private boolean enabled;

    /**
     * The PvPCWorld constructor.
     * @param level - The level instance.
     */
    public PvPCWorld(Level level)
    {
        this.knockbackInfo = new PvPCKnockback();
        this.enabled = true;
        this.level = level;
        this.levelName = level.getName();
    }

    /**
     * The PvPCWorld constructor.
     * @param level - The level name.
     * @param enabled - Determines whether the world is enabled.
     * @param knockback - The knockback object.
     */
    public PvPCWorld(String level, boolean enabled, PvPCKnockback knockback)
    {
        this.knockbackInfo = knockback;
        this.levelName = level;
        this.level = Server.getInstance().getLevelByName(level);
        this.enabled = enabled;
    }

    /**
     * Gets the level name of the world.
     * @return - The level name.
     */
    public String getLevelName()
    {
        return this.levelName;
    }

    /**
     * Sets the level of the world, mainly used for updates.
     * @param level - The new level.
     */
    public void setLevel(Level level)
    {
        this.levelName = level.getName();
        this.level = level;
    }

    /**
     * Gets the level that this world belongs to.
     * @return - The level.
     */
    public Level getLevel()
    {
        return this.level;
    }

    /**
     * Determines whether the world's knockback is enabled.
     * @return - True if enabled, false otherwise.
     */
    public boolean isKBEnabled()
    {
        return this.enabled;
    }

    /**
     * Sets the kb as enabled or not.
     * @param enabled - The value that determines whether it's enabled.
     */
    public void setKBEnabled(boolean enabled)
    {
        this.enabled = enabled;
    }

    /**
     * Gets the knockback information.
     * @return - The knockback info.
     */
    public PvPCKnockback getKnockback()
    {
        return this.knockbackInfo;
    }

    /**
     * Determines if each player can use the knockback.
     * @param player - The player object.
     * @param attacker - The player who is attacking the player.
     * @return - true if the players can use the knockback.
     */
    @Override
    public boolean canUseKnockback(Player player, Player attacker)
    {
        // TODO: Determine if the player can use the knockback.
        return false;
    }

    /**
     * Exports the value to a HashMap.
     * @return The hashmap containing the data.
     */
    @Override
    public HashMap<String, Object> export()
    {
        HashMap<String, Object> data = new HashMap<>();
        data.put("kbInfo", this.knockbackInfo.export());
        data.put("kbEnabled", this.enabled);
        return data;
    }

    /**
     * Determines if another object is equal to another object.
     * @param object - The object we are comparing.
     * @return - true if equivalent, false otherwise.
     */
    public boolean equals(Object object)
    {
        if(object instanceof PvPCWorld)
        {
            Level level = ((PvPCWorld) object).level;
            if(level != null && this.level != null)
            {
                return Utils.levelsEqual(this.level, level);
            }
        }

        return false;
    }

    /**
     * Decodes the object into a PvPCWorld object, the new format.
     * @param levelName - The corresponding level name.
     * @param object - The encoded object.
     * @return - A new PvPCWorld instance if successful, null otherwise.
     */
    public static PvPCWorld decode(String levelName, Object object)
    {
        if(
                object instanceof JSONObject
                && ((JSONObject) object).containsKey("kbEnabled")
                && ((JSONObject) object).containsKey("kbInfo")
        )
        {
            PvPCKnockback knockback = PvPCKnockback.decode(((JSONObject) object).get("kbInfo"));
            Object enabled = ((JSONObject) object).get("kbEnabled");

            if(knockback != null && enabled instanceof Boolean)
            {
                return new PvPCWorld(
                        levelName,
                        (boolean)enabled,
                        knockback
                );
            }
        }
        return null;
    }
}
